src/py-opentimelineio/opentimelineio-bindings/otio_serializableObjects.cpp

```diff
@@ -51,6 +52,47 @@ namespace {
             return py::str(thing);
         }
     }
+
+    template<typename T, typename U>
+    bool each_child(T* t, py::object descended_from_type, optional<TimeRange> const& search_range, bool shallow_search, std::vector<SerializableObject*>& l) {
+        if (descended_from_type.is(py::type::handle_of<U>()))
+        {
+            for (const auto& child : t->template each_child<U>(ErrorStatusHandler(), search_range, shallow_search)) {
+                l.push_back(child.value);
+            }
+            return true;
+        }
+        return false;
+    }
+
+    template<typename T>
+    std::vector<SerializableObject*> each_child(T* t, py::object descended_from_type, optional<TimeRange> const& search_range, bool shallow_search = false) {
+        std::vector<SerializableObject*> l;
+        if (each_child<T, Clip>(t, descended_from_type, search_range, shallow_search, l)) ;
+        else if (each_child<T, Composition>(t, descended_from_type, search_range, shallow_search, l)) ;
+        else if (each_child<T, Gap>(t, descended_from_type, search_range, shallow_search, l)) ;
+        else if (each_child<T, Item>(t, descended_from_type, search_range, shallow_search, l)) ;
+        else if (each_child<T, Stack>(t, descended_from_type, search_range, shallow_search, l)) ;
+        else if (each_child<T, Timeline>(t, descended_from_type, search_range, shallow_search, l)) ;
+        else if (each_child<T, Track>(t, descended_from_type, search_range, shallow_search, l)) ;
+        else if (each_child<T, Transition>(t, descended_from_type, search_range, shallow_search, l)) ;
+        else
+        {
+            for (const auto& child : t->template each_child<Composable>(ErrorStatusHandler(), search_range, shallow_search)) {
+                l.push_back(child.value);
+            }
+        }
+        return l;
+    }
+    
+    template<typename T>
+    std::vector<SerializableObject*> each_clip(T* t, optional<TimeRange> const& search_range, bool shallow_search = false) {
+        std::vector<SerializableObject*> l;
+        for (const auto& child : t->each_clip(ErrorStatusHandler(), search_range, shallow_search)) {
+            l.push_back(child.value);
+        }
+        return l;
```

Comment:
The subtle distinction here from the original implementation is that it returns a static list rather than an iterator?  Meaning, it has to build the entire list in memory and then return that, rather than using the `yield` style iterator?

Reply:
Yes, AFAIK there's not an equivalent to ```yield``` in C++. Since the vector is just holding pointers to the objects it shouldn't be too expensive, unless you anticipate large sizes like over 10K. The speed of the function might also be faster in C++ though I haven't done any benchmarks comparing the two implementations.

If that changes the style of the Python usage too much we could bring back the Python implementation, though it seemed nice to share the code. Another possibility would be sort of a hybrid approach where instead of Python code calling the C++ wrapper function directly, there was a Python version of ```each_child()/each_clip()``` that used the C++ functions but exposed a more Python like interface.

Reply:
@reinecke what do you think?  Is it ok to give up the iterator-interface of `each_clip()`?  I kind of feel like the name is misleading if it isn't an iterator.  Something like `all_clips()` is more intuitive to me if it builds a list.

Reply:
Something like this might work as a hybrid approach?
```
def each_clip(self, search_range=None, descended_from_type=_otio.Composable):
    # _each_clip() is the C++ implementation
    for child in _each_clip(search_range, descended_from_type):
        yield child
```

Reply:
The c++ equivalent of yield is any of the container accessors that return an iterator, since an iterator is somewhat equivalent to a generator. The big difference is that a generator is going to hold a reference on a container so the container won't vanish under foot, but a C++ iterator does not hold a reference. Due to the complexity of the OTIO data structures, an iterator would be complex to implement and fragile in the sense that the iterator has to be very cognizant of implementation details. Returning a vector makes more sense for C++ makes a lot more sense to avoid the complexity and fragility.

It seems to me that you don't have to give up the public facing part of the implementation, as a python generator could be wrap the list from the C++ implementation. The generator would be responsible for cleaning up the vector when the vector is exhausted, since the magic thing about generators is you can only iterate them forward, once.

So I think we can have the (assumed) efficiency of the C++ implementation, but still return a generator for iteration. @ssteinbach since you're saying the `yield` style iterator, I have read that as `generator`. Hopefully I haven't read too much into that.

Reply:
I'm not sure if you looked into using pybind's make_iterator but it might be helpful to you.  There are some examples here:
https://github.com/pybind/pybind11/blob/master/tests/test_sequences_and_iterators.cpp

If you really wanted to add yield-like functionality on the c++ side as well, you'd likely have to add your own iterator class: 
https://www.cplusplus.com/reference/iterator/iterator/ 

That would probably be a fair amount of work.  A less involved solution might be to use make_iterator on the returned vector.  That way at least the generation of the Python instances from the c++ ones would be generated.

Reply:
Hi Roger, thanks for the links! I did not know about the pybind11 iterators, I will check them out. Do you think that's a better approach than the small Python code snippet I posted above?

Reply:
No problem, I hope they help!

I think one thing that is nice about the make_iterator approach is that you can add this at the end (as the examples do):
```
py::keep_alive<0, 1>()
```
So you effectively bind the lifetime of the generator and the vector (in your case) together and ensuring the vector only needs to be build once.  By calling the c++ method from python each time, I think it would generate the vector on each call, even though the python objects are generated, right?  

I haven't tried this approach myself, so I might be wrong about it.  It might not be worth the effort if you try it and it starts crashing everywhere because of the vector's lifetime.

Reply:
I tried adding ```py::make_iterator``` to the returned vector but unfortunately got seg-faults in the tests again, which I assume is because the vector is being created on the stack and getting destroyed prematurely. It seems like you would need a custom container class to get it to work right.

I also tried the hybrid Python approach and it seemed to work OK, and didn't re-generate the vector on each call. Which makes sense since otherwise how would it be able to iterate? Python must keep the state of the function around for the lifespan of the iterator and "yields" control flow back to the calling function. However one downside of this approach is that each Python class (Timeline, Stack, Track, etc.) needs the function definitions which is a fair bit of boiler-plate code.

So, to sum up it seems like there are three options:

1. Leave changes as-is. Pros: probably the least amount of code of the three options. Cons: maybe not "Pythonic" without the iterators?
2. Use the "hybrid" approach. Pros: fairly simple and keeps the "Pythonic" iterators. Cons: a bit of Python boiler-plate code as I mentioned above.
3. Use a custom container class with "py::make_iterator". Pros: probably the most "correct" way of the three options. Cons: probably adds the most code of the three options...

Reply:
The fourth option is to keep the implementation of `each_child` and `each_clip` for the python implementation _in_ python, duplicating the code in each place.

Reply:
True, but that duplicates quite a bit of code, which isn't that complex but isn't trivial either. I think that also implies adding C++ confidence tests to verify that the C++ versions of the functions work the same way as the Python versions.

I think the "hybrid" approach is functionally the same, it leaves the existing Python API as-is but replaces the implementations with the C++ functions. Maybe it was wrong to call it boiler-plate code in that respect.

Reply:
My personal preference (C++ biased) would be for option 1, since it is the least amount of code and anything in Python inheriting from ```Composition``` will automatically get support for ```each_child()```. Though if the Python iterators are important to keep I like option 2, since it is also a relatively small amount of code and just needs some additional Python to decorate each class to support ```each_child()```.